src/snowflake/cli/plugins/nativeapp/v2_conversions/v2_to_v1_decorator.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from functools import wraps
+from typing import Any, Dict, Optional
+
+from click import ClickException
+from snowflake.cli.api.cli_global_context import cli_context, cli_context_manager
+from snowflake.cli.api.project.schemas.entities.application_entity import (
+    ApplicationEntity,
+)
+from snowflake.cli.api.project.schemas.entities.application_package_entity import (
+    ApplicationPackageEntity,
+)
+from snowflake.cli.api.project.schemas.project_definition import (
+    DefinitionV11,
+    DefinitionV20,
+)
+
+
+def _pdf_v2_to_v1(v2_definition: DefinitionV20) -> DefinitionV11:
+    pdfv1: Dict[str, Any] = {"definition_version": "1.1", "native_app": {}}
+
+    app_package_definition: ApplicationPackageEntity = None
```

Comment:
```suggestion
    app_package_definition: Optional[ApplicationPackageEntity] = None
```

Reply:
There must be an app package entity, otherwise an exception will be raised.
The type checker is not smart enough to narrow the type after `if not app_package_definition: raise`, so I put the "final" type here. Not sure if there's a more conventional mypy way to do it -- open for suggestions if so.

Reply:
You could invert the if clause to get this narrowing working, but it's not important enough to change the flow of the function to something that's (IMO) worse. I prefer the way you've done it, where we make the happy path the one that goes to the end of the function.